juju/juju
# Caas uniter

## Description of change

The PR reowrks the caas operator worker to remove its responsibility to run charm hooks. Instead, it will start individual uniter workers for each unit that is added. The PR is broken up into several commits outlined below.

Note - there's a bunch of now superceded code which is not yet deleted. That cleanup will be done in a followup PR. Also needing followup is status and presence for caas application/unit agents.

Commit 1 - state changes.
The state.Unit behaviour depends on the model type, in particular the ShouldBeAssigned() helper. CAAS units are not assigned to a machine. Also, caas unit agents now have status. Still TODO is work to properly manage agent presence.

Commit 2 - api facade changes
The caas operator facades gain some new methods to watch units and get the life of an entity. Methods to get amd watch charm config are deleted. CAAS models are need to access the uniter and eadership facades.

The other main change is that now uniter, leadership, hook retry strategy (et al) facades need to work where the authenticated connection is for an application rather than a unit). The uniter workers started by the caas operator all get given an application api connection (auth tag is the application) so tweaks were needed to the auth checking logic to allow this.

Commit 3 - caas operator worker
Starts uniter workers instead of running hooks itself

Commit 4 - uniter worker
Various tweaks to account for the fact that caas units do not have storage, nor machine public/private addresses. Nor does a caas uniter install the charm - this is done by the caas operator. The set relation status call also now takes an explicit unit tag rather than relying on the auth connection.

Commit 5 - manifold changes to start the caas operator with its revised dependencies

## QA steps

Deploy kubernetes core (validates IAAS models)
Deploy a CAAS model and gitlab application.
Add units etc.
Check status-log to see that expected hooks fire.
